code/datums/components/food/decomposition.dm

```diff
@@ -0,0 +1,162 @@
+//"Don't leave food on the floor, that's how we get ants"
+
+#define DECOMPOSITION_TIME 10 MINUTES
+#define DECOMPOSITION_TIME_RAW 5 MINUTES
+#define DECOMPOSITION_TIME_GROSS 7 MINUTES
+
+/datum/component/decomposition
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+	/// Makes sure food only starts decomposing if a player's EVER picked it up before
+	var/handled = FALSE
+	/// Used to stop food in someone's hand & in storage slots from decomposing.
+	var/protected = FALSE
+	/// Used to stop the timer & check for the examine proc
+	var/timerid
+	/// Used so the timer won't reset.
+	var/time_remaining = DECOMPOSITION_TIME
+	/// Used to give raw/gross food lower timers
+	var/decomp_flags
+	/// Used for examining
+	var/examine_type = 0
+
+/datum/component/decomposition/Initialize(decomp_flags = NONE)
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	src.decomp_flags = decomp_flags
+
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/handle_movement)
+	RegisterSignal(parent, list(
+		COMSIG_ITEM_PICKUP, //person picks up an item
+		COMSIG_STORAGE_ENTERED), //Object enters a storage object (boxes, etc.)
+		.proc/picked_up)
+	RegisterSignal(parent, list(
+		COMSIG_ITEM_DROPPED, //Object is dropped anywhere
+		COMSIG_STORAGE_EXITED), //Object exits a storage object (boxes, etc)
+		.proc/dropped)
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/examine) // Self-explanitory
+
+	if(decomp_flags & GROSS && !(decomp_flags & RAW))
+		time_remaining = DECOMPOSITION_TIME_GROSS
+		examine_type = 1
+	if(decomp_flags & RAW) // Raw food overrides gross
+		time_remaining = DECOMPOSITION_TIME_RAW
+		examine_type = 2
+
+/datum/component/decomposition/UnregisterFromParent()
+	UnregisterSignal(parent, list(
+		COMSIG_ITEM_PICKUP,
+		COMSIG_STORAGE_ENTERED,
+		COMSIG_MOVABLE_MOVED,
+		COMSIG_ITEM_DROPPED,
+		COMSIG_STORAGE_EXITED,
+		COMSIG_PARENT_EXAMINE))
+
+/datum/component/decomposition/proc/handle_movement() // Assuming it is guaranteed to be food
+	var/obj/item/food/food = parent
+	var/atom/last_loc = food.loc
+
+	var/clean = FALSE // Used to check if it's on a clean surface
+	while(last_loc && !isarea(last_loc))
+		if(HAS_TRAIT(food.loc, TRAIT_PROTECT_FOOD))
+			clean = TRUE
+			break
+		else if(locate(/obj/structure/table) in food.loc)
+			clean = TRUE
+			break
+		last_loc = last_loc.loc
+	if(!handled || clean || protected)
+		// prevent decomposition
+		if(active_timers)
```

Comment:
Why check for active_timers? Could you use the timer id  itself as an indicator? so if the timer's fired or if it's deleted the id is null, otherwise it's an actual  number?

Reply:
If there's no active timer then changing time_remaining & the deltimer proc both have runtimes, so I wanted to make sure that didn't happen. I can try to check for timerid if you think it'll work better, but it functions fine at the moment

Reply:
good enough for me